### PR TITLE
Fix some shellcheck failures in hack

### PR DESCRIPTION
**Pull request by SataQiu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Fix some shellcheck failures in hack :
- hack/list-feature-tests.sh
- hack/verify-govet.sh
- hack/verify-staging-meta-files.sh

**Which issue(s) this PR fixes**:
Ref #72956

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by spiffxp** [comment]
taking a look on behalf of @kubernetes/bash-firefighters

**Review by spiffxp — COMMENTED** [review]
One question

File: hack/verify-govet.sh [hunk 1]
```diff
@@ -20,12 +20,6 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
 
-# For help output
-ARGHELP=""
-if [[ "$#" -gt 0 ]]; then
-    ARGHELP="WHAT='$@'"
```
[the remark this answers is not in the document]
Comment (cblecker) [reply hack/verify-govet.sh]: Yeah, I don't think this is safe to remove.

File: hack/verify-no-vendor-cycles.sh [hunk 2]
```diff
@@ -18,14 +18,15 @@ set -o errexit
...
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
 
-staging_repos=($(ls "${KUBE_ROOT}/staging/src/k8s.io/"))
+IFS=" " read -r -a staging_repos <<< "$(ls "${KUBE_ROOT}/staging/src/k8s.io/")"
 staging_repos_pattern=$(IFS="|"; echo "${staging_repos[*]}")
 
 failed=false
-for i in $(find vendor/ -type d); do
-  deps=$(go list -f '{{range .Deps}}{{.}}{{"\n"}}{{end}}' ./$i 2> /dev/null || echo "")
+while IFS= read -r -d '' i
```
Comment (cblecker) [inline hack/verify-no-vendor-cycles.sh]: concerned about this change in logic. would love to see this tested to ensure it works as intended.

File: hack/verify-govet.sh [hunk 3]
```diff
@@ -20,12 +20,13 @@ set -o errexit
...
 set -o pipefail
 
-KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
 
 # For help output
 ARGHELP=""
 if [[ "$#" -gt 0 ]]; then
-    ARGHELP="WHAT='$@'"
+    # shellcheck disable=2034 # Variables sourced in other scripts executed from the same shell
```
Comment (spiffxp) [inline hack/verify-govet.sh]: I still feel like I need more explanation here.  Why is this necessary? What's using ARGHELP and why do we have to do it this way vs. some other way that doesn't trigger shellcheck?

File: hack/verify-staging-meta-files.sh [hunk 4]
```diff
@@ -18,9 +18,9 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
 
-staging_repos=($(ls "${KUBE_ROOT}/staging/src/k8s.io/"))
+IFS=" " read -r -a staging_repos <<< "$(ls "${KUBE_ROOT}/staging/src/k8s.io/")"
```
Comment (spiffxp) [inline hack/verify-staging-meta-files.sh]: which shellcheck failure is this working around?

Comment (SataQiu) [reply hack/verify-staging-meta-files.sh]: This is the check output of `hack/verify-shellcheck.sh` :
```
In ./hack/verify-staging-meta-files.sh line 23:
staging_repos=($(ls "${KUBE_ROOT}/staging/src/k8s.io/"))
               ^-- SC2207: Prefer mapfile or read -a to split command output (or quote to avoid splitting).
```

Comment (BenTheElder) [reply hack/verify-staging-meta-files.sh]: `read` is more explicit and portable than `(<a delimited string>)`. we can't use `mapfile` due to macOS

File: hack/verify-govet.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenTheElder) [reply hack/verify-govet.sh]: what do we think is consuming this?

[the remark this answers is not in the document]
Comment (SataQiu) [reply hack/verify-govet.sh]: @BenTheElder 
I adopted a new way to solve this shellcheck problem.
Just add some `echo` output.
Refer to 
`hack/verify-all.sh` :
```
# For help output
ARGHELP=""
if [[ -n "${KUBE_VERIFY_GIT_BRANCH:-}" ]]; then
    ARGHELP="BRANCH=${KUBE_VERIFY_GIT_BRANCH}"
fi

echo "NOTE: $0 has been replaced by 'make verify'"
echo
echo "The equivalent of this invocation is: "
echo "    make verify ${ARGHELP}"
echo
echo
make --no-print-directory -C "${KUBE_ROOT}" verify BRANCH="${KUBE_VERIFY_GIT_BRANCH:-}"
```
`hack/build-go.sh`
```
# For help output
ARGHELP=""
if [[ "$#" -gt 0 ]]; then
    ARGHELP="WHAT='$*'"
fi

echo "NOTE: $0 has been replaced by 'make' or 'make all'"
echo
echo "The equivalent of this invocation is: "
echo "    make ${ARGHELP}"
echo
echo
make --no-print-directory -C "${KUBE_ROOT}" all WHAT="$*"
```

File: hack/verify-staging-meta-files.sh [hunk 5]
```diff
@@ -36,8 +36,8 @@ exceptions=(
 )
 
 RESULT=0
-for repo in ${staging_repos[@]}; do
-  for filename in ${expected_filenames[@]}; do
+for repo in "${staging_repos[@]}"; do
```
Comment (spiffxp) [inline hack/verify-staging-meta-files.sh]: FWIW I saw a much less complicated looking approach for getting "the list of staging repos" in https://github.com/kubernetes/kubernetes/pull/74332, is that insufficient?

File: hack/verify-staging-meta-files.sh (hunk 4 above)
Comment (spiffxp) [reply hack/verify-staging-meta-files.sh]: I'm not completely clear why we need to read this in as an array, see my comment below

File: hack/verify-staging-meta-files.sh (hunk 5 above)
Comment (ixdy) [reply hack/verify-staging-meta-files.sh]: using quotes here is wrong, since it prevents splitting (which we want)

Comment (SataQiu) [reply hack/verify-staging-meta-files.sh]: @ixdy 
I've tested whether quoting changes the behavior of the script. The result is NO!
First Script (without quotes):
```
#!/bin/bash

expected_filenames=(
  .github/PULL_REQUEST_TEMPLATE.md
  code-of-conduct.md
  LICENSE
  OWNERS
  README.md
  SECURITY_CONTACTS
)

for filename in ${expected_filenames[@]}; do
	echo $filename
done
=====
OUTPUT:
.github/PULL_REQUEST_TEMPLATE.md
code-of-conduct.md
LICENSE
OWNERS
README.md
SECURITY_CONTACTS
```
First Script (with quotes):
```
#!/bin/bash

expected_filenames=(
  .github/PULL_REQUEST_TEMPLATE.md
  code-of-conduct.md
  LICENSE
  OWNERS
  README.md
  SECURITY_CONTACTS
)

for filename in "${expected_filenames[@]}"; do
	echo $filename
done
OUTPUT:
.github/PULL_REQUEST_TEMPLATE.md
code-of-conduct.md
LICENSE
OWNERS
README.md
SECURITY_CONTACTS

```

Comment (SataQiu) [reply hack/verify-staging-meta-files.sh]: Double quotes around $@ (and similarly, ${array[@]}) prevents globbing and word splitting of individual elements, while still expanding to multiple separate arguments.

Let's say you have three arguments: baz, foo bar and *

"$@" will expand into exactly that: baz, foo bar and *

$@ will expand into multiple other arguments: baz, foo, bar, file.txt and otherfile.jpg

Since the latter is rarely expected or desired, ShellCheck warns about it.

Ref: https://github.com/koalaman/shellcheck/wiki/SC2068

Comment (ixdy) [reply hack/verify-staging-meta-files.sh]: you're right, I think I'm confusing this with the case where it isn't already an array.

File: hack/verify-staging-meta-files.sh [hunk 6]
```diff
@@ -36,8 +34,8 @@ exceptions=(
 )
 
 RESULT=0
-for repo in ${staging_repos[@]}; do
-  for filename in ${expected_filenames[@]}; do
+for repo in ${KUBE_ROOT}/staging/src/k8s.io/*; do
```
Comment (ixdy) [inline hack/verify-staging-meta-files.sh]: I think this is not quite equivalent.
```console
$ staging_repos=($(ls "staging/src/k8s.io/"))
$ echo ${staging_repos[@]}
api apiextensions-apiserver apimachinery apiserver client-go cli-runtime cloud-provider cluster-bootstrap code-generator component-base csi-api csi-translation-lib kube-aggregator kube-controller-manager kubelet kube-proxy kube-scheduler metrics node-api sample-apiserver sample-cli-plugin sample-controller
```
vs
```
$ for repo in staging/src/k8s.io/*; do echo $repo; done
staging/src/k8s.io/api
staging/src/k8s.io/apiextensions-apiserver
staging/src/k8s.io/apimachinery
staging/src/k8s.io/apiserver
staging/src/k8s.io/client-go
staging/src/k8s.io/cli-runtime
staging/src/k8s.io/cloud-provider
staging/src/k8s.io/cluster-bootstrap
staging/src/k8s.io/code-generator
staging/src/k8s.io/component-base
staging/src/k8s.io/csi-api
staging/src/k8s.io/csi-translation-lib
staging/src/k8s.io/kube-aggregator
staging/src/k8s.io/kube-controller-manager
staging/src/k8s.io/kubelet
staging/src/k8s.io/kube-proxy
staging/src/k8s.io/kube-scheduler
staging/src/k8s.io/metrics
staging/src/k8s.io/node-api
staging/src/k8s.io/sample-apiserver
staging/src/k8s.io/sample-cli-plugin
staging/src/k8s.io/sample-controller
```

(note the prefix added when using the glob)

File: hack/verify-staging-meta-files.sh (hunk 4 above)
Comment (BenTheElder) [reply hack/verify-staging-meta-files.sh]: that's fair, but that is the previous behavior which imo was reasonable enough, as bash goes :^)
unless we're redesigning entirely, `($( ... ))` -> `read -r ...` is an improvement

File: hack/verify-staging-meta-files.sh (hunk 6 above)
Comment (SataQiu) [reply hack/verify-staging-meta-files.sh]: @ixdy You're right! It is not quite equivalent. I changed it back!

**Comment by SataQiu** [comment]
@spiffxp @ixdy @cblecker 
Thank you all for your review.
But this PR has been around for a long time. 
Do I still need to do something more?

**Comment by spiffxp** [comment]
I think I’m suggesting we use the simpler globing approach in the PR I linked unless someone sees something wrong with that

File: hack/verify-govet.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (spiffxp) [reply hack/verify-govet.sh]: ok, following build-go makes sense to me
